app/models/manageiq/providers/storage_manager/cinder_manager/refresh_helper_methods.rb

```diff
@@ -0,0 +1,173 @@
+      :type              => "ManageIQ::Providers::Openstack::CloudManager::CloudVolume",
+      :name              => volume_name(volume),
+      :status            => volume.status,
+      :bootable          => volume.attributes['bootable'],
+      :creation_time     => volume.created_at,
+      :description       => volume_description(volume),
+      :volume_type       => volume.volume_type,
+      :snapshot_uid      => volume.snapshot_id,
+      :size              => volume.size.to_i.gigabytes,
+      :tenant            => @data_index.fetch_path(:cloud_tenants, volume.tenant_id),
+      :availability_zone => @data_index.fetch_path(:availability_zones, "volume-" + volume.availability_zone || "null_az"),
+    }
+
+    volume.attachments.each do |a|
+      if a['device'].blank?
+        _log.warn "#{log_header}: Volume: #{uid}, is missing a mountpoint, skipping the volume processing"
+        _log.warn "#{log_header}:   EMS: #{@ems.name}, Instance: #{a['server_id']}"
+        next
+      end
+
+      dev = File.basename(a['device'])
+      disks = @data_index.fetch_path(:vms, a['server_id'], :hardware, :disks)
+
+      unless disks
+        _log.warn "#{log_header}: Volume: #{uid}, attached to instance not visible in the scope of this EMS"
+        _log.warn "#{log_header}:   EMS: #{@ems.name}, Instance: #{a['server_id']}"
+        next
+      end
+
+      if (disk = disks.detect { |d| d[:location] == dev })
+        disk[:size] = new_result[:size]
+      else
+        disk = add_instance_disk(disks, new_result[:size], dev, "OpenStack Volume")
+      end
+
+      if disk
+        disk[:backing]      = new_result
+        disk[:backing_type] = 'CloudVolume'
+      end
+    end
+
+    return uid, new_result
+  end
+
+  def volume_name(volume)
+    # Cinder v1
+    return volume.display_name if volume.respond_to?(:display_name)
+    # Cinder v2
+    return volume.name
+  end 
+
+  def volume_description(volume)
+    # Cinder v1
+    return volume.display_description if volume.respond_to?(:display_description)
+    # Cinder v2
+    return volume.description
+  end 
+
+  def add_instance_disk(disks, size, location, name, controller_type)
+    if size >= 0
+      disk = {
+        :device_name     => name,
+        :device_type     => "disk",
+        :controller_type => controller_type,
+        :location        => location,
+        :size            => size
+      }
+      disks << disk
+      return disk
+    end
+    nil
+  end
+
+  def link_storage_associations
+    @data[:cloud_volumes].each do |cv|
+      #
+      # Associations between volumes and the snapshots on which
+      # they are based, if any.
+      #
+      base_snapshot_uid = cv.delete(:snapshot_uid)
+      base_snapshot = @data_index.fetch_path(:cloud_volume_snapshots, base_snapshot_uid)
+      cv[:base_snapshot] = base_snapshot unless base_snapshot.nil?
+    end if @data[:cloud_volumes]
+  end
+
+  def process_collection(collection, key, &block)
```

Comment:
@hsong-rh From here down, are helper methods. The methods above this are parsing methods.